server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java

```diff
@@ -264,4 +264,12 @@ protected Set<Long> getSnapshotIdsOnlyInPrimaryStorage(long volumeId) {
 
         return snapshotIdsOnlyInPrimaryStorage;
     }
+
+    public SnapshotInfo convertSnapshotIfNeeded(SnapshotInfo snapshotInfo) {
+        if (snapshotInfo.getParent() == null) {
```

Comment:
@JoaoJandre, if a snapshot is copied to a second zone and create a template from it on that zone when you try to deploy a VM it fails because the qemu attempts to access the secondary storage of the first zone

> [qemu-img convert -O qcow2 -U --image-opts driver=qcow2,file.filename=/mnt/b5fc4e51-bfea-36d8-b2db-557f84a9f353/aa60ba30-a8cc-44d7-a603-709a50bac890.qcow2 --bitmaps /mnt/b206833c-4a77-3b14-9de5-3ac91cc6c417/ba4e0f05-12e3-4d10-b051-62cf5264c70f ] encountered the error: [qemu-img: Could not open 'driver=qcow2,file.filename=/mnt/b5fc4e51-bfea-36d8-b2db-557f84a9f353/aa60ba30-a8cc-44d7-a603-709a50bac890.qcow2': Could not open backing file: Could not open '/mnt/dc75bca1-2951-366c-88a2-bb4c840b0148/snapshots/2/994/9f53b6b3-4018-4c45-a82b-73f30dc570aa': No such file or directory].

I'm still exploring this feature but my guess for this issue is that the copies on the second zone don't have parent ID and aren't converted.

Reply:
@slavkap this should be fixed now, could you test it again?

Reply:
@JoaoJandre, all works good! But isn't it better to transfer the parent IDs to the copied snapshots? My point is that when you copy the snapshot it increases the time if there is a need for the snapshot to be converted, and this could be avoided only in cases when you need to create a template/volume from the snapshot on the second zone

Reply:
@slavkap  In the future, we could copy the snapshot chain to the other secondary storage. However, further changes would need to be done, for example, we would need to rebase the snapshots with their new parents (otherwise, we run into the exception you mentioned above), that means changing more processes on the SSVM, also we would need to add `qemu-img` to the System VM template. 
Since the current implementation is working, and this PR is already pretty big, I would rather introduce this feature as is and optimize it in the future if needed.